### PR TITLE
Now that's what I call a tokenizer! Vol. 9000

### LINKED ISSUE #17: Create tokenizer.
https://webassembly.github.io/spec/core/text/lexical.html#tokens

### PR DESCRIPTION
**Pull request by jasonwyatt**
Sorry for the sheer size of this PR. We can talk through it tomorrow morning if you like!

Fixes #17

### CODE REVIEW

File: library/src/main/java/kwasm/format/text/Name.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@
 package kwasm.format.text
 
 import kwasm.format.ParseContext
+import kwasm.format.text.token.StringLiteral
```
Comment (welnaseth) [inline library/src/main/java/kwasm/format/text/Name.kt]: Is this import needed? there are no other changes in the file

File: library/src/main/java/kwasm/format/text/Tokenizer.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    /** Tokenizes source code from the provided [source] [String]. */
+    fun tokenize(
+        source: String,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> = tokenize(StringReader(source), context)
+
+    // TODO: it might be worth considering making a lot of these functions `inline fun`s.
+    private fun buildActualTokensFrom(token: RawToken): List<Token> = when {
+        token.sequence.isEmpty() -> emptyList()
+        Format.PATTERN.get().matchEntire(token.sequence) != null -> emptyList()
```
Comment (welnaseth) [inline library/src/main/java/kwasm/format/text/Tokenizer.kt]: What does this match?

File: library/src/main/java/kwasm/format/text/Tokenizer.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        token.isKeyword() -> listOf(token.toKeyword())
+        token.isIntegerLiteral() -> listOf(token.toIntegerLiteral())
+        token.isFloatLiteral() -> listOf(token.toFloatLiteral())
+        token.isStringLiteral() -> listOf(token.toStringLiteral())
+        token.isIdentifier() -> listOf(token.toIdentifier())
+        token.isOpenParen() || token.isClosedParen() -> listOf(token.toParen())
+        token.isReserved() -> listOf(token.toReserved())
+        else -> {
```
Comment (welnaseth) [inline library/src/main/java/kwasm/format/text/Tokenizer.kt]: So when parsing the raw token, if it could parse the token to exactly one type, it will be set? Otherwise this else figures out the longest match?

File: library/src/main/java/kwasm/format/text/Tokenizer.kt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                token.findParen(),
+                token.findReserved()
+            ).maxBy { it?.sequence?.length ?: -1 }.takeIf { it != null }
+                ?: throw ParseException(
+                    "No valid token found in sequence: \"${token.sequence}\"",
+                    token.context
+                )
+
+            // Recursively get token list from the contents before the longest found token.
+            val prefixTokens = buildActualTokensFrom(
```
Comment (welnaseth) [inline library/src/main/java/kwasm/format/text/Tokenizer.kt]: Could you explain this again, what cases do we need to recursively get tokens?

File: library/src/main/java/kwasm/format/text/Type.kt [hunk 5]
```diff
@@ -18,6 +18,7 @@ import kwasm.ast.Limit
 import kwasm.format.ParseContext
 import kwasm.format.ParseException
 import kwasm.format.shiftColumnBy
+import kwasm.format.text.token.IntegerLiteral
```
Comment (welnaseth) [inline library/src/main/java/kwasm/format/text/Type.kt]: No changes refrence this import, is it needed?

File: library/src/main/java/kwasm/format/text/Name.kt (hunk 1 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Name.kt]: Yeah, because StringLiteral got moved

File: library/src/main/java/kwasm/format/text/Tokenizer.kt (hunk 2 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Tokenizer.kt]: I'll add a comment - it matches format character strings... eg spaces, newlines, tabs

File: library/src/main/java/kwasm/format/text/Tokenizer.kt (hunk 3 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Tokenizer.kt]: Yep!

File: library/src/main/java/kwasm/format/text/Tokenizer.kt (hunk 4 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Tokenizer.kt]: An example could be:  `"this is a string"+1.5e10aKeyword`

The longest found token in this would be `"this is a string"`, but we still need to get `+1.5e10` and `aKeyword` out.

File: library/src/main/java/kwasm/ast/Limit.kt [hunk 6]
```diff
@@ -14,10 +14,9 @@
...
 
-import kwasm.format.text.IntegerLiteral
 import kwasm.format.text.Type.Limits
 
 /**
  * This is a data class that represents the limits of a [Limits] object.
  */
-data class Limit(val min: IntegerLiteral.Unsigned, val max: IntegerLiteral.Unsigned)
\ No newline at end of file
+data class Limit(val min: ULong, val max: ULong)
```
Comment (welnanick) [inline library/src/main/java/kwasm/ast/Limit.kt]: You can undo these changes. They got covered in my PR. Just revert this file to origin/master

Comment (welnanick) [inline library/src/main/java/kwasm/ast/Limit.kt]: Revert this file to origin/master, my PR covered these changes but will definitely cause merge conflicts.

File: library/src/main/java/kwasm/format/text/Type.kt (hunk 5 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Type.kt]: Similar to `Name` - this is because IntegerLiteral was moved

File: library/src/main/java/kwasm/ast/Limit.kt (hunk 6 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/ast/Limit.kt]: 👍

File: library/src/main/java/kwasm/format/text/Type.kt [hunk 7]
```diff
@@ -115,7 +116,7 @@ sealed class Type<T>(
                     throw ParseException("Invalid number of arguments. Expected 1 or 2 but found 0", context)
                 }
                 val min = IntegerLiteral.Unsigned(sequence, 32, context)
-                Limit(min, IntegerLiteral.Unsigned(UInt.MAX_VALUE.toString(), 32, null))
+                Limit(min.value, UInt.MAX_VALUE.toULong())
```
Comment (welnanick) [inline library/src/main/java/kwasm/format/text/Type.kt]: Covered in my PR

File: library/src/main/java/kwasm/format/text/Type.kt [hunk 8]
```diff
@@ -130,7 +131,7 @@ sealed class Type<T>(
                     // We must undo the context shift if we encounter this error
                     throw ParseException("Invalid Range specified, min > max. Found min: $min, max: $max", context)
                 }
-                Limit(min, max)
+                Limit(min.value, max.value)
```
Comment (welnanick) [inline library/src/main/java/kwasm/format/text/Type.kt]: Covered in my PR

File: library/src/test/java/kwasm/format/text/LimitsTest.kt [hunk 9]
```diff
@@ -156,4 +156,4 @@ class LimitsTest {
             .isInstanceOf(ParseException::class.java)
             .hasMessageContaining("Invalid number of arguments.")
     }
-}
\ No newline at end of file
+}
```
Comment (welnanick) [inline library/src/test/java/kwasm/format/text/LimitsTest.kt]: Covered by my PR

File: library/src/main/java/kwasm/format/text/Tokenizer.kt [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    ): List<Token> {
+        val rawTokens = source.use { Comment.stripComments(it.readText(), context) }.tokens
+        return rawTokens.flatMap(::buildActualTokensFrom)
+    }
+
+    /** Tokenizes source code from the provided [source] [String]. */
+    fun tokenize(
+        source: String,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> = tokenize(StringReader(source), context)
```
Comment (welnanick) [inline library/src/main/java/kwasm/format/text/Tokenizer.kt]: Won't we want to strip comments here as well?

File: library/src/main/java/kwasm/ast/Limit.kt (hunk 6 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/ast/Limit.kt]: Done

File: library/src/main/java/kwasm/format/text/Type.kt (hunk 7 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Type.kt]: Done

File: library/src/main/java/kwasm/format/text/Type.kt (hunk 8 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Type.kt]: Done

File: library/src/main/java/kwasm/format/text/Tokenizer.kt (hunk 10 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/format/text/Tokenizer.kt]: yeah, it's calling the other function - which will strip the comments

File: library/src/main/java/kwasm/ast/Limit.kt [hunk 11]
```diff
@@ -14,8 +14,6 @@
 
 package kwasm.ast
 
-/**
- * This is a data class that represents the limits of a Limits object.
- */
```
Comment (welnaseth) [inline library/src/main/java/kwasm/ast/Limit.kt]: Do we want to delete the comment?

File: library/src/main/java/kwasm/format/text/Tokenizer.kt (hunk 4 above)
Comment (welnaseth) [reply library/src/main/java/kwasm/format/text/Tokenizer.kt]: I see

File: library/src/main/java/kwasm/ast/Limit.kt (hunk 11 above)
Comment (jasonwyatt) [reply library/src/main/java/kwasm/ast/Limit.kt]: I'll add it back
